integration-cli/docker_cli_attach_unix_test.go

```diff
@@ -91,32 +91,52 @@ func TestAttachAfterDetach(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Could not open pty: %v", err)
 	}
+	defer cpty.Close()
 
 	cmd = exec.Command(dockerBinary, "attach", name)
 	cmd.Stdin = tty
 	cmd.Stdout = tty
 	cmd.Stderr = tty
 
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+
+	cmdErr := make(chan error)
 	go func() {
-		if err := cmd.Run(); err != nil {
-			t.Fatalf("attach returned error %s", err)
-		}
-		cpty.Close() // unblocks the reader in case of a failure
+		cmdErr <- cmd.Wait()
 	}()
 
-	time.Sleep(500 * time.Millisecond)
-	cpty.Write([]byte("\n"))
-	time.Sleep(500 * time.Millisecond)
+	readErr := make(chan error)
 	bytes := make([]byte, 10)
+	var nBytes int
+	go func() {
+		time.Sleep(500 * time.Millisecond)
+		cpty.Write([]byte("\n"))
+		time.Sleep(500 * time.Millisecond)
 
-	n, err := cpty.Read(bytes)
+		nBytes, err = cpty.Read(bytes)
+		readErr <- err
+	}()
 
-	if err != nil {
-		t.Fatalf("prompt read failed: %v", err)
+	select {
+	case err := <-readErr:
```

Comment:
No, I mean that your test is not deterministic, because you can get signal through `readErr` or `cmdErr` and in branch about `cmdErr` you read `readErr` anyway, so why we need to wait on `readErr` concurrently with `cmdErr` if process is still expected to finish.